src/kernl/model_optimization.py

```diff
@@ -12,46 +12,50 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 #
-from typing import Callable, List
+from typing import List
 
 import torch
-import torchdynamo
+import torch._dynamo as torchdynamo
 from transformers import PreTrainedModel
 
 from kernl.implementations.cuda_graph import cuda_graphs_wrapper
 from kernl.optimizer.dynamo_backend import dynamo_backend_ofi
 
 
-def optimize_model(original_model: PreTrainedModel, pool: (int, int) = torch.cuda.graph_pool_handle()) -> Callable:
+# single shared pool by default
+_pool: (int, int) = torch.cuda.graph_pool_handle()
+
+
+# needs to be generated once to be reused several times, like encoder/decoder models
+# https://github.com/pytorch/torchdynamo/issues/1816
+def _compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+    dynamo_backend_ofi(gm)
+    return cuda_graphs_wrapper(gm, example_inputs, pool=_pool)
+
+
+def optimize_model(original_model: PreTrainedModel) -> None:
     """
     Optimizes a given model. Optimization is done in two steps:
     *  first step is to convert the given model to fx graph.
     *  second step is to replace patterns found in the graph in order to optimize the model.
+    Example:
```

Comment:
To generate doc we will probably need to change this in the future

Reply:
Yes, but there is probably more than that to modify :-)